imagebuildah/chroot_symlink.go

```diff
@@ -0,0 +1,115 @@
+)
+
+func init() {
+	reexec.Register(symlinkChrootedCommand, resolveChrootedSymlinks)
+}
+
+func resolveChrootedSymlinks() {
+	status := 0
+	flag.Parse()
+	if len(flag.Args()) < 1 {
+		os.Exit(1)
+	}
+	// Our first parameter is the directory to chroot into.
+	if err := unix.Chdir(flag.Arg(0)); err != nil {
+		fmt.Fprintf(os.Stderr, "chdir(): %v", err)
+		os.Exit(1)
+	}
+	if err := unix.Chroot(flag.Arg(0)); err != nil {
+		fmt.Fprintf(os.Stderr, "chroot(): %v", err)
+		os.Exit(1)
+	}
+
+	// Our second paramter is the mountpoint/rootdir of the container
+	symLink, err := getSymbolicLink(flag.Arg(0), flag.Arg(1))
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "error getting symbolic links: %v", err)
+		os.Exit(1)
+	}
+	if _, err := os.Stdout.WriteString(symLink); err != nil {
+		fmt.Fprintf(os.Stderr, "error writing string to stdout: %v", err)
+		os.Exit(1)
+	}
+	os.Exit(status)
+}
+
+func resolveSymLink(rootdir, filename string) (string, error) {
+	// The child process expects a chroot and one path that
+	// will be consulted relative to the chroot directory and evaluated
+	// for any symbolic links present.
+	cmd := reexec.Command(symlinkChrootedCommand, rootdir, filename)
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return "", errors.Wrapf(err, string(output))
+	}
+
+	// Hand back the resolved symlink, will be "" if a symlink is not found
+	return string(output), nil
+}
+
+func getSymbolicLink(rootdir, path string) (string, error) {
+	var (
+		symPath          string
+		symLinksResolved int
+	)
+
+	splitPath := strings.Split(path, "/")
+	if splitPath[0] == "" {
+		splitPath = splitPath[1:]
+		symPath = "/"
+	}
+
+	for _, p := range splitPath {
+		// If we have resolved 40 symlinks, that means something is terribly wrong
+		// will return an error and exit
+		if symLinksResolved > 39 {
+			return "", errors.Errorf("have resolved 40 symlinks, something is terribly wrong!")
+		}
+		symPath = filepath.Join(symPath, p)
+		info, err := os.Lstat(symPath)
+		if os.IsNotExist(err) {
+			if err = os.MkdirAll(symPath, 0755); err != nil {
+				return "", errors.Wrapf(err, "error ensuring volume path %q exists", symPath)
+			}
+			info, err = os.Lstat(symPath)
+		}
+		// Move on to the next part of the path is if this part is not a symlink
+		if info.Mode()&os.ModeSymlink != os.ModeSymlink {
+			continue
+		}
+
+		// Read the symlink to get what it points to
+		targetDir, err := os.Readlink(symPath)
+		if err != nil {
+			return "", errors.Wrapf(err, "error reading link %q", symPath)
+		}
+		// if symlink points to itself, return an error
+		if targetDir == symPath {
+			return "", errors.Errorf("symlink %q points to itself: %q", symPath, targetDir)
+		}
+		// if the symlink points to a relative path, prepend the path till now to the resolved path
+		if !filepath.IsAbs(targetDir) {
+			targetDir = filepath.Join(symPath, targetDir)
+		}
+		// run filepath.Clean to remove the ".." from relative paths
+		symPath = filepath.Clean(targetDir)
```

Comment:
At this point, `symPath` could now be resolved to a new pathname which is also, itself, a symbolic link (and on, and on).  Do we handle this?

Reply:
Just added functionality to handle this. PTAL